crates/ra_syntax/src/grammar/expressions/atom.rs

```diff
@@ -427,12 +430,20 @@ fn continue_expr(p: &mut Parser) -> CompletedMarker {
 //         break 'l 92;
 //     }
 // }
-fn break_expr(p: &mut Parser) -> CompletedMarker {
+fn break_expr(p: &mut Parser, r: Restrictions) -> CompletedMarker {
     assert!(p.at(BREAK_KW));
     let m = p.start();
     p.bump();
     p.eat(LIFETIME);
-    if p.at_ts(EXPR_FIRST) {
+    // test break_ambiguity
+    // fn foo(){
+    //     if break {}
+    //     while break {}
+    //     for i in break {}
+    //     match break {}
+    // }
+    if r.forbid_structs && p.at_ts(EXPR_FIRST_NO_BLOCK) || !r.forbid_structs && p.at_ts(EXPR_FIRST)
+    {
         expr(p);
```

Comment:
Should we pass `r` down to `expr` here?

Is `if break S {} {}` parsed correctly?

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=55e91f597b049520bb4b7047dbe136f6 seems to suggest it's allowed, and we do allow it currently. I think we're alright. I might rename the field on Restrictions though to be more descriptive.